pkg/component/kubernetes/apiserver/horizontalpodautoscaler.go

```diff
@@ -74,6 +86,83 @@ func (k *kubeAPIServer) reconcileHorizontalPodAutoscaler(ctx context.Context, hp
 
 		return nil
 	})
+	return err
+}
+
+func (k *kubeAPIServer) reconcileHorizontalPodAutoscalerInVPAAndHPAMode(ctx context.Context, hpa *autoscalingv2.HorizontalPodAutoscaler, deployment *appsv1.Deployment) error {
+	// The chosen value is 6 CPU: 1 CPU less than the VPA's maxAllowed 7 CPU in VPAAndHPA mode to have a headroom for the horizontal scaling.
+	hpaTargetAverageValueCPU := resource.MustParse("6")
+	// The chosen value is 24G: 4G less than the VPA's maxAllowed 28G in VPAAndHPA mode to have a headroom for the horizontal scaling.
+	hpaTargetAverageValueMemory := resource.MustParse("24G")
+
+	_, err := controllerutils.GetAndCreateOrMergePatch(ctx, k.client.Client(), hpa, func() error {
+		minReplicas := k.values.Autoscaling.MinReplicas
+		if k.values.Autoscaling.ScaleDownDisabled && hpa.Spec.MinReplicas != nil {
+			// If scale-down is disabled and the HPA resource exists and HPA's spec.minReplicas is not nil,
+			// then minReplicas is max(spec.minReplicas, status.desiredReplcias).
+			// When scale-down is disabled, this allows operators to specify a custom value for HPA spec.minReplicas
```

Comment:
> When scale-down is disabled, this allows operators to specify a custom value for HPA spec.minReplicas

Is this a new thing? My understanding of the current behavior was that when `alpha.control-plane.scaling.shoot.gardener.cloud/scale-down-disabled` was set, the replicas were fixe to `4`?

If that is still the case, I guess we can just also make it a static `4` here, wdyt?

Reply:
Yes, it is something new that @vlerenc suggested to add. When the HPA exists and the scale-down-disabled annotation is added to a Shoot, we will no longer set minReplicas=maxReplicas=4 but we will keep the current HPA minReplicas (.status.desiredReplicas). IIUC, this is closer to the the definition of the `alpha.control-plane.scaling.shoot.gardener.cloud/scale-down-disabled` annotation - previously when this annotation is set we were doing scale up to 4 replicas; now we make sure that we don't do scale down below minReplicas of the existing HPA.
The logic should also allow operator to modify the HPA's minReplicas count manually (if needed) when the scale-down annotation is present.

So, currently when the scale-down-disabled annotation is set the minReplicas would be `4` only when the HPA resources does not exists - new Shoot clusters. Otherwise, minReplicas will be always `max(*hpa.Spec.MinReplicas, hpa.Status.DesiredReplicas)`.

Reply:
Yes, indeed. After learning of this annotation, I found it somewhat intuitive/intuitively named as it pretends to disable scale-down while in fact fixing the number of replicas to 4. I think, what we want in ops is more freedom to set custom values. If I want 3 for what has normally a minimum of 2, I set minReplicas to 3. On the next reconciliation, my decision as operator is respected (it doesn't have to be always 4). If we allow for up to 6 replicas later (maybe we will) and I as an operator want 3 or 5, it will work as well. 4 is a historic choice (because that's how the KAPI HVPA resources were set), but it feels strange (to stick to it now that HVPA will be gone) without this historic knowledge and it feels wrong even with it (why 4). Snowflaking should not fix the replicas, but avoid touching what we operators configure (within tolerated limits). That said, I think we may later further improve our snowflaking capabilities, but for now, removing this odd number (that will eventually bite us/surprise people not in the know) seems an appropriate course of action as we remove the very reason (HVPA) that influenced the decision for 4 in the first place.

Of course you can say, why touch it? Sure, but above is my reasoning why I would. The number feels very wrong and I like to forget about this number "4" going forward and make any manual snowflaking changes (as an operator) more predicable and less magic.

Reply:
Technically, shouldn't we also set `maxReplicas` to be equal to `minReplicas`, even with the current proposal? 
If I set the `minReplicas` in the HPA to 3, `maxReplicas` would still be 4. Depending on when reconciliation happens and when `desiredReplicas` changes there could still be scale down from 4 to 3 replicas even if the annotaiton is set:

1. Let's say that `desiredReplicas` is 3, then this code sets `minReplicas` to 3 and `maxReplicas` to 4.
2. Replicas get scaled up to 4.
3. Then, no reconciliation happens to set `minReplicas` to the new number of `desiredReplicas` (4) for some time.
4. Some time later (before the next reconciliation occurs), scale down happens and replicas are set back to 3.

Reply:
Since I'd like to change max to 6 eventually anyway (there is not reason to have a time-bomb like this in the code, should clusters grow even more over time), I would not do that (force min=max). If the cluster is snowflaked and someone decided for min=3 then so be it. If it scales up and down, it will do so, but not below 3. If someone decided for 4, 5, or 6, then that's it. If someone decided for min=max, then that's it. The operator decides.

Besides, I have the hope, that the snowflaking was only necessary, because HVPA was terrible.

Reply:
> At the very least: We should document how this annotation works, how you can interact (modify the HPA's `minReplicas`, not the Deployment's `replicas`) and note that horizontal downscaling _does_ happen.

Agree. I tried to document it in https://github.com/gardener/gardener/pull/9678/commits/1433c5f02a925019dfb441eee33fea43d592ab3e.